src/thor/map_matcher.cc

```diff
@@ -129,63 +130,88 @@ MapMatcher::FormPath(meili::MapMatcher* matcher,
                      const std::shared_ptr<sif::DynamicCost>* mode_costing,
                      const sif::TravelMode mode,
                      std::vector<std::pair<GraphId, GraphId>>& disconnected_edges,
-                     const Options options) {
+                     Options& options) {
   // Set costing based on the mode
   const auto& costing = mode_costing[static_cast<uint32_t>(mode)];
   bool use_timestamps = options.use_timestamps();
-
   // Iterate through the matched path. Form PathInfo - populate elapsed time
   // Return an empty path (or throw exception) if path is not connected.
   Cost elapsed;
   std::vector<PathInfo> path;
   GraphId prior_edge, prior_node;
-  const NodeInfo* nodeinfo = nullptr;
-  const DirectedEdge* directededge;
   EdgeLabel pred;
+  const GraphTile* tile = nullptr;
+  const DirectedEdge* directededge = nullptr;
+  const NodeInfo* nodeinfo = nullptr;
+
+  // We support either the epoch timestamp that came with the trace point or
+  // a local date time which we convert to epoch by finding the first timezone
+  uint32_t origin_epoch = 0;
+  for (const auto& s : edge_segments) {
+    if (!s.edgeid.Is_Valid() || !matcher->graphreader().GetGraphTile(s.edgeid, tile))
+      continue;
+    directededge = tile->directededge(s.edgeid);
+    if (matcher->graphreader().GetGraphTile(directededge->endnode(), tile)) {
+      // get the timezone
+      nodeinfo = tile->node(directededge->endnode());
+      const auto* tz = DateTime::get_tz_db().from_index(nodeinfo->timezone());
+      // if its timestamp based need to signal that out to trip leg builder
+      if (options.shape(0).time() != -1.0) {
```

Comment:
yeah, maybe to avoid confusion it should take the api::Location as an argument. the reason being is that this is gnerated code so the only way to actually do this would be with a free function which of course leads to some ambiguity imho